review-batch: single-review JSON failure aborts entire session

## Problem

During the scheduled review run ([run #25707852006](https://github.com/petry-projects/.github-private/actions/runs/25707852006)), the `single-review` step invoked `claude-opus-4-7` to confirm an approval for [PR #129](https://github.com/petry-projects/.github-private/pull/129). The model ran for ~74 seconds and returned a non-JSON response, causing the entire batch session to abort — skipping all 35 remaining PR candidates.

### Failure trace
```
[approve] triage cleared — running single confirmation (claude-opus-4-7)
##[error]single-review did not produce valid JSON
##[error]Review failed for .../pull/129 (exit code 1)
##[error]Session aborted early after failure on .../pull/129 (exit code 1). Skipped 35 remaining candidate(s)
Summary: 0 reviews posted, 0 no-ops skipped, 1 failures (processed 1/36 candidates) [SESSION ABORTED EARLY]
```

### Root cause
The audit/confirmation model (`claude-opus-4-7`) did not return structured JSON — likely due to a verbose non-JSON response, a refusal, or a timeout/context-limit issue. The `scripts/review-batch.sh` script treats any per-PR failure as fatal and exits, aborting all remaining candidates.

## Acceptance Criteria

- [ ] **Resilient JSON parsing:** If `single-review` (or any review engine call) returns non-JSON output, the error is caught, logged with the raw output, and treated as a per-PR failure — not a fatal crash.
- [ ] **Per-PR isolation:** A failure on one PR does not abort the session. Remaining candidates continue to be processed.
- [ ] **Retry logic:** The `single-review` step retries at least once (with a short delay) before marking the PR as failed, to handle transient model errors.
- [ ] **Fallback behavior:** If the audit model fails after retries, the system can optionally fall back to the triage-level decision (or skip and flag for manual review) rather than hard-failing.
- [ ] **Accurate summary:** The final session summary correctly reports the number of skipped/failed PRs and does not claim `SESSION ABORTED EARLY` when only individual PRs fail.
- [ ] **No regression:** Existing review logic (triage → escalation → deep review → audit flow) continues to work correctly for PRs that do not encounter errors.

--- a/scripts/review-batch.sh
+++ b/scripts/review-batch.sh
@@ -103,13 +103,10 @@ while IFS= read -r pr_url; do
       abort_reason="rate-limit on fallback engine"
       ;;
     *)
-      # Other failure — session-fatal. A systemic problem (model degraded,
-      # prompt regression, malformed verdicts) shouldn't burn the queue.
+      # Per-PR failure — log and continue. Remaining candidates still run.
+      # Accumulated failures surface in the session summary.
       failed=$((failed + 1))
       echo "::error::Review failed for $pr_url (exit code $rc)"
-      session_aborted=1
-      abort_pr="$pr_url"
-      abort_reason="exit code $rc"
       ;;
   esac
 

--- a/scripts/review-one-pr.sh
+++ b/scripts/review-one-pr.sh
@@ -325,14 +325,30 @@ if [ "$TRIAGE_ESCALATE" = "false" ]; then
     rm -f "$VERDICT_JSON" "$VERDICT_JSON.raw"
     OUTPUT_FILE="$VERDICT_JSON"
     export OUTPUT_FILE
-    if run_agentic prompts/single-review.md "$ENGINE_SINGLE_MODEL" \
-         > "$VERDICT_JSON.raw" 2>"/tmp/cascade/single-review.log" \
-       && extract_verdict_json "$VERDICT_JSON.raw" "$VERDICT_JSON"; then
+    SINGLE_LOG="/tmp/cascade/single-review-attempt-${single_attempt}.log"
+
+    single_rc=0
+    run_agentic prompts/single-review.md "$ENGINE_SINGLE_MODEL" \
+      > "$VERDICT_JSON.raw" 2>"$SINGLE_LOG" || single_rc=$?
+
+    # Check for rate limit before retrying — exit code 2 lets review-batch.sh
+    # trigger engine fallback rather than burning retries on the same engine.
+    SINGLE_STDOUT=$(cat "$VERDICT_JSON.raw" 2>/dev/null || true)
+    SINGLE_STDERR=$(cat "$SINGLE_LOG" 2>/dev/null || true)
+    if is_rate_limited "$SINGLE_STDOUT" || is_rate_limited "$SINGLE_STDERR"; then
+      echo "    [approve] rate limit detected — exiting with code 2 for engine fallback"
+      [ -n "$SINGLE_STDERR" ] && echo "    limit stderr: $SINGLE_STDERR"
+      exit 2
+    fi
+
+    if [ "$single_rc" -eq 0 ] && extract_verdict_json "$VERDICT_JSON.raw" "$VERDICT_JSON"; then
       single_ok=true
       break
     fi
-    echo "    [approve] single-review attempt $single_attempt/$SINGLE_REVIEW_MAX_RETRIES produced no valid JSON"
+
+    echo "    [approve] single-review attempt $single_attempt/$SINGLE_REVIEW_MAX_RETRIES produced no valid JSON (exit $single_rc)"
     head -20 "$VERDICT_JSON.raw" 2>/dev/null || true
+    [ -n "$SINGLE_STDERR" ] && echo "    stderr: $SINGLE_STDERR"
     if [ "$single_attempt" -lt "$SINGLE_REVIEW_MAX_RETRIES" ]; then
       echo "    [approve] retrying in ${SINGLE_REVIEW_RETRY_DELAY_SEC}s"
       sleep "$SINGLE_REVIEW_RETRY_DELAY_SEC"
@@ -342,7 +358,10 @@ if [ "$TRIAGE_ESCALATE" = "false" ]; then
 
   if [ "$single_ok" = "false" ]; then
     echo "::warning::single-review failed after $SINGLE_REVIEW_MAX_RETRIES attempts — flagging for manual review"
-    [ -s "/tmp/cascade/single-review.log" ] && cat "/tmp/cascade/single-review.log"
+    for i in $(seq 1 "$SINGLE_REVIEW_MAX_RETRIES"); do
+      log="/tmp/cascade/single-review-attempt-${i}.log"
+      [ -s "$log" ] && { echo "    attempt $i stderr:"; cat "$log"; }
+    done
     if [ "${DRY_RUN:-false}" = "false" ]; then
       gh pr edit "$PR_URL" --add-label needs-human-review 2>/dev/null || true
     fi

--- a/scripts/review-one-pr.sh
+++ b/scripts/review-one-pr.sh
@@ -316,10 +316,38 @@ if [ "$TRIAGE_ESCALATE" = "false" ]; then
   echo "    [approve] triage cleared — running single confirmation ($ENGINE_SINGLE_MODEL)"
   export REVIEW_MODE="triage-approved"
   VERDICT_JSON="/tmp/cascade/single-review-verdict.json"
-  OUTPUT_FILE="$VERDICT_JSON"
-  export OUTPUT_FILE
-  run_agentic prompts/single-review.md "$ENGINE_SINGLE_MODEL" > "$VERDICT_JSON.raw"
-  extract_verdict_json "$VERDICT_JSON.raw" "$VERDICT_JSON" || { echo "::error::single-review did not produce valid JSON"; exit 1; }
+  SINGLE_REVIEW_MAX_RETRIES="${SINGLE_REVIEW_MAX_RETRIES:-2}"
+  SINGLE_REVIEW_RETRY_DELAY_SEC="${SINGLE_REVIEW_RETRY_DELAY_SEC:-15}"
+
+  single_attempt=1
+  single_ok=false
+  while [ "$single_attempt" -le "$SINGLE_REVIEW_MAX_RETRIES" ]; do
+    rm -f "$VERDICT_JSON" "$VERDICT_JSON.raw"
+    OUTPUT_FILE="$VERDICT_JSON"
+    export OUTPUT_FILE
+    if run_agentic prompts/single-review.md "$ENGINE_SINGLE_MODEL" \
+         > "$VERDICT_JSON.raw" 2>"/tmp/cascade/single-review.log" \
+       && extract_verdict_json "$VERDICT_JSON.raw" "$VERDICT_JSON"; then
+      single_ok=true
+      break
+    fi
+    echo "    [approve] single-review attempt $single_attempt/$SINGLE_REVIEW_MAX_RETRIES produced no valid JSON"
+    head -20 "$VERDICT_JSON.raw" 2>/dev/null || true
+    if [ "$single_attempt" -lt "$SINGLE_REVIEW_MAX_RETRIES" ]; then
+      echo "    [approve] retrying in ${SINGLE_REVIEW_RETRY_DELAY_SEC}s"
+      sleep "$SINGLE_REVIEW_RETRY_DELAY_SEC"
+    fi
+    single_attempt=$((single_attempt + 1))
+  done
+
+  if [ "$single_ok" = "false" ]; then
+    echo "::warning::single-review failed after $SINGLE_REVIEW_MAX_RETRIES attempts — flagging for manual review"
+    [ -s "/tmp/cascade/single-review.log" ] && cat "/tmp/cascade/single-review.log"
+    if [ "${DRY_RUN:-false}" = "false" ]; then
+      gh pr edit "$PR_URL" --add-label needs-human-review 2>/dev/null || true
+    fi
+    exit 1
+  fi
   rm -f "$VERDICT_JSON.raw"
 
   # Post the review using the verdict

